### PR TITLE
Make dependencies which require native extensions optional

### PR DESCRIPTION
**Pull request by jimbo**
## Description

Some project dependencies have native extensions written in other languages. These dependencies require [`node-gyp`](https://github.com/nodejs/node-gyp#installation), which in turn requires Python and a C/C++ compiler (msbuild.exe on Windows, gcc or clang on *nixes, etc).

Since these dependencies are required, any user or environment without these tools, such as a minimal Docker container or a fresh Windows/OSX install, will fail to install the project dependencies.

The affected dependencies are:
- `shrink-ray-current`, a middleware for compressing HTTP responses, used by `upward-js`
- `sharp`, a library for image compression and manipulation, used by the image optimization middleware in `pwa-buildpack`
- `bundlesize`, a tool for reporting changes in build asset filesize to a GitHub integration reporter

Since none of the above features are strictly critical for basic operations, this PR moves them to `optionalDependencies`, so that npm/yarn will not bail out when it can't build them.

When these optional dependencies are not installed:
- In a CI container which runs `yarn run bundlesize`, that step will fail. CircleCI NodeJS build containers do not appear to have this problem.
- In a development or production environment running the NodeJS server, HTTP responses and catalog images won't be compressed. In dev mode, this will affect performance measurements negatively. In production mode, it's not acceptable; some external proxy would have to do compression instead.

## Related Issue

#820

## How Has This Been Tested?

`yarn test`

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have linked an issue to this PR.
- [x] I have indicated the change type and relevant package(s).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All CI checks are green (linting, build/deploy, etc).
- [ ] At least one core contributor has approved this PR.

### CODE REVIEW

**Comment by rowan-m** [comment]
Maybe I'm just doing something wrong here, but:
```
rowan@penguin:~/src/pwa-studio$ git branch 
* jimbo/optional-deps
  release/2.0

rowan@penguin:~/src/pwa-studio$ yarn install --ignore-optional
yarn install v1.12.3
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.4: The platform "linux" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > jest-transform-graphql@2.1.0" has unmet peer dependency "graphql-tag@^2.0.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "@babel/runtime@~7.2.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-dom@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-router-dom@~4.4.0-beta.6".
warning " > @magento/upward-js@2.0.0-rc.19" has unmet peer dependency "express@~4.16.4".
warning "workspace-aggregator-56d8621d-e949-4ce6-84bc-aba50a343aad > @magento/peregrine > @storybook/react@4.1.6" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-56d8621d-e949-4ce6-84bc-aba50a343aad > @magento/pwa-buildpack > @magento/express-sharp@2.1.1-newdeps.1" has unmet peer dependency "sharp@^0.21.2".
warning "workspace-aggregator-56d8621d-e949-4ce6-84bc-aba50a343aad > @magento/venia-concept > react-feather@1.1.5" has unmet peer dependency "prop-types@>= 15".
warning "workspace-aggregator-56d8621d-e949-4ce6-84bc-aba50a343aad > @magento/peregrine > @storybook/react > @storybook/core@4.1.6" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-56d8621d-e949-4ce6-84bc-aba50a343aad > @magento/peregrine > @storybook/react > @emotion/styled > @emotion/styled-base@0.10.6" has unmet peer dependency "@emotion/core@0.x.x".
[5/5] Building fresh packages...
[1/5] ⠁ iltorb
[-/5] ⠁ waiting...
[-/5] ⠁ waiting...
[4/5] ⠁ sharp
error /home/rowan/src/pwa-studio/node_modules/iltorb: Command failed.
Exit code: 1
Command: detect-libc prebuild-install || node-gyp rebuild
Arguments: 
Directory: /home/rowan/src/pwa-studio/node_modules/iltorb
Output:
prebuild-install WARN install No prebuilt binaries found (target=10.14.2 runtime=node arch=x64 platform=linux)
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@10.14.2 | linux | x64
gyp ERR! configure error 
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (/home/rowan/src/pwa-studio/node_modules/node-gyp/lib/configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (/home/rowan/src/pwa-studio/node_modules/node-gyp/lib/configure.js:406:16)
gyp ERR! stack     at F (/home/rowan/src/pwa-studio/node_modules/which/which.js:68:16)
gyp ERR! stack     at E (/home/rowan/src/pwa-studio/node_modules/which/which.js:80:29)
gyp ERR! stack     at /home/rowan/src/pwa-studio/node_modules/which/which.js:89:16
gyp ERR! stack     at /home/rowan/src/pwa-studio/node_modules/isexe/index.js:42:5
gyp ERR! stack     at /home/rowan/src/pwa-studio/node_modules/isexe/mode.js:8:5
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:154:21)
gyp ERR! System Linux 4.19.16-02893-g2cf2c17c8a43
gyp ERR! command "/home/rowan/.nvm/versions/node/v10.14.2/bin/node" "/home/rowan/src/pwa-studio/node_modules/iltorb/node_modules/.bin/node-gyp" "rebuild"
gyp ERR! cwd /home/rowan/src/pwa-studio/node_modules/iltorb

rowan@penguin:~/src/pwa-studio$ yarn why node-gyp
yarn why v1.12.3
[1/4] Why do we have the module "node-gyp"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "node-gyp@3.8.0"
info Reasons this module exists
   - "_project_#@magento#upward-js#shrink-ray-current#node-zopfli-es" depends on it
   - Hoisted from "_project_#@magento#upward-js#shrink-ray-current#node-zopfli-es#node-gyp"
   - Hoisted from "_project_#bundlesize#brotli-size#iltorb#node-gyp"
   - Hoisted from "_project_#lerna#libnpm#npm-lifecycle#node-gyp"
info Disk size without dependencies: "2.31MB"
info Disk size with unique dependencies: "3.55MB"
info Disk size with transitive dependencies: "10.38MB"
info Number of shared dependencies: 58
Done in 1.29s.
```

**Comment by tjwiebell** [comment]
@jimbo 

Looks like `node-gyp` is also a dependency of `express-sharp`, which was added to `pwa-buildpack`.

https://github.com/magento-research/pwa-studio/pull/768/files#diff-85fd7ec16271bfcd3b7a870b232a7915R36

**Comment by sirugh** [comment]
Sadness. On my windows 10 computer after checking out this branch:

```bash
C:\Users\Stephen\code\pwa-studio>yarn install
yarn install v1.13.0
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.4: The platform "win32" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > jest-transform-graphql@2.1.0" has unmet peer dependency "graphql-tag@^2.0.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "@babel/runtime@~7.2.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-dom@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-router-dom@~4.4.0-beta.6".
warning " > @magento/upward-js@2.0.0-rc.19" has unmet peer dependency "express@~4.16.4".
warning "workspace-aggregator-c7c61cc6-c5fb-44af-8ffc-54e322aa8896 > @magento/peregrine > @storybook/react@4.1.6" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-c7c61cc6-c5fb-44af-8ffc-54e322aa8896 > @magento/pwa-buildpack > @magento/express-sharp@2.1.1-newdeps.1" has unmet peer dependency "sharp@^0.21.2".
warning "workspace-aggregator-c7c61cc6-c5fb-44af-8ffc-54e322aa8896 > @magento/venia-concept > react-feather@1.1.5" has unmet peer dependency "prop-types@>= 15".
warning "workspace-aggregator-c7c61cc6-c5fb-44af-8ffc-54e322aa8896 > @magento/peregrine > @storybook/react > @storybook/core@4.1.6" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-c7c61cc6-c5fb-44af-8ffc-54e322aa8896 > @magento/peregrine > @storybook/react > @emotion/styled > @emotion/styled-base@0.10.6" has unmet peer dependency "@emotion/core@0.x.x".
[5/5] Building fresh packages...
[1/7] ⠠ iltorb
[7/7] ⠠ webpack-cli
[3/7] ⠠ iltorb
[4/7] ⠠ node-zopfli-es
warning Error running install script for optional dependency: "C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@10.15.1 | win32 | x64
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable \"python\", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:509:16)
gyp ERR! stack     at C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\graceful-fs\\polyfills.js:282:31
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:154:21)
gyp ERR! System Windows_NT 10.0.17763
gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\Stephen\\\\code\\\\pwa-studio\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"rebuild\"
gyp ERR! cwd C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es
[1/7] ⠁ iltorb
[-/7] ⠁ waiting...
[-/7] ⠁ waiting...
[-/7] ⠁ waiting...
error C:\Users\Stephen\code\pwa-studio\node_modules\iltorb: Command failed.
Exit code: 1
Command: detect-libc prebuild-install || node-gyp rebuild
Arguments:
Directory: C:\Users\Stephen\code\pwa-studio\node_modules\iltorb
Output:
prebuild-install WARN install No prebuilt binaries found (target=10.15.1 runtime=node arch=x64 platform=win32)
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@10.15.1 | win32 | x64
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable "python", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\Users\Stephen\code\pwa-studio\node_modules\node-gyp\lib\configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\Users\Stephen\code\pwa-studio\node_modules\node-gyp\lib\configure.js:509:16)
gyp ERR! stack     at C:\Users\Stephen\code\pwa-studio\node_modules\graceful-fs\polyfills.js:282:31
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:154:21)
gyp ERR! System Windows_NT 10.0.17763
gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
gyp ERR! cwd C:\Users\Stephen\code\pwa-studio\node_modules\iltorb
```

**Comment by zetlen** [comment]
Easy fix. We just gotta add every binary dep to optionalDependencies, and make sure that requiring them is behind a try-catch. I can do that now and update your brach, @jimbo

**Comment by zetlen** [comment]
This is ready for any of the @core-pwa-devs or the reporters of #820 and related bugs to test on their various Windows and *nix setups:
@henktheunissen
@dpatil-magento 
@sharkySharks

**Comment by sirugh** [comment]
Success? Got the warning for installing of optional dep which is fine.

```
Stephen@Stephen-PC MINGW64 ~/code/pwa-studio ((b4a008eb...))
$ yarn install
yarn install v1.13.0
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
info fsevents@1.2.4: The platform "win32" is incompatible with this module.
info "fsevents@1.2.4" is an optional dependency and failed compatibility check. Excluding it from installation.
[4/5] Linking dependencies...
warning " > jest-transform-graphql@2.1.0" has unmet peer dependency "graphql-tag@^2.0.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "@babel/runtime@~7.2.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-dom@~16.8.0".
warning " > @magento/peregrine@2.0.0-rc.19" has unmet peer dependency "react-router-dom@~4.4.0-beta.6".
warning " > @magento/upward-js@2.0.0-rc.19" has unmet peer dependency "express@~4.16.4".
warning "workspace-aggregator-80c460ae-1a82-43ce-927f-5f644c21263b > @magento/peregrine > @storybook/react@4.1.11" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-80c460ae-1a82-43ce-927f-5f644c21263b > @magento/pwa-buildpack > @magento/express-sharp@2.1.1-newdeps.1" has unmet peer dependency "sharp@^0.21.2".
warning "workspace-aggregator-80c460ae-1a82-43ce-927f-5f644c21263b > @magento/venia-concept > react-feather@1.1.5" has unmet peer dependency "prop-types@>= 15".
warning "workspace-aggregator-80c460ae-1a82-43ce-927f-5f644c21263b > @magento/peregrine > @storybook/react > @storybook/core@4.1.11" has unmet peer dependency "babel-loader@^7.0.0 || ^8.0.0".
warning "workspace-aggregator-80c460ae-1a82-43ce-927f-5f644c21263b > @magento/peregrine > @storybook/react > @emotion/styled > @emotion/styled-base@0.10.6" has unmet peer dependency "@emotion/core@0.x.x".
[5/5] Building fresh packages...
[7/7] ⠠ webpack-cli
[2/7] ⠠ iltorb
[3/7] ⠠ iltorb
[4/7] ⠠ node-zopfli-es
warning Error running install script for optional dependency: "C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es: Command failed.
Exit code: 1
Command: node-gyp rebuild
Arguments:
Directory: C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es
Output:
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@10.15.1 | win32 | x64
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable \"python\", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:509:16)
gyp ERR! stack     at C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\graceful-fs\\polyfills.js:282:31
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:154:21)
gyp ERR! System Windows_NT 10.0.17763
gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\Stephen\\\\code\\\\pwa-studio\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"rebuild\"
gyp ERR! cwd C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-zopfli-es
[-/7] ⠁ waiting...
[2/7] ⠁ iltorb
[-/7] ⠁ waiting...
[-/7] ⠂ waiting...
warning Error running install script for optional dependency: "C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\iltorb: Command failed.
Exit code: 1
Command: detect-libc prebuild-install || node-gyp rebuild
Arguments:
Directory: C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\iltorb
Output:
prebuild-install WARN install No prebuilt binaries found (target=10.15.1 runtime=node arch=x64 platform=win32)
gyp info it worked if it ends with ok
gyp info using node-gyp@3.8.0
gyp info using node@10.15.1 | win32 | x64
gyp ERR! configure error
gyp ERR! stack Error: Can't find Python executable \"python\", you can set the PYTHON env variable.
gyp ERR! stack     at PythonFinder.failNoPython (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:484:19)
gyp ERR! stack     at PythonFinder.<anonymous> (C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\node-gyp\\lib\\configure.js:509:16)
gyp ERR! stack     at C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\graceful-fs\\polyfills.js:282:31
gyp ERR! stack     at FSReqWrap.oncomplete (fs.js:154:21)
gyp ERR! System Windows_NT 10.0.17763
gyp ERR! command \"C:\\\\Program Files\\\\nodejs\\\\node.exe\" \"C:\\\\Users\\\\Stephen\\\\code\\\\pwa-studio\\\\node_modules\\\\node-gyp\\\\bin\\\\node-gyp.js\" \"rebuild\"
gyp ERR! cwd C:\\Users\\Stephen\\code\\pwa-studio\\node_modules\\iltorb
Done in 29.07s.
```

File: package.json [hunk 1]
```diff
@@ -18,7 +18,6 @@
   "scripts": {
     "build": "yarn workspaces run build",
     "build:dev": "yarn workspaces run build:dev",
-    "bundlesize": "npx bundlesize",
```
Comment (sharkySharks) [inline package.json]: do we want to remove this script? We still want to run it if it isn't optional right?

Comment (zetlen) [reply package.json]: Yarn has a feature where [you can run any locally installed CLI with `yarn <cli>` without defining a script.](https://yarnpkg.com/lang/en/docs/cli/#locally-installed-clis) So we don't need this script defined, and now it's not cluttering package.json.

**Comment by zetlen** [comment]
Thanks @sharkySharks ! Still waiting on some Windows people.

**Comment by jimbo** [comment]
@zetlen Thanks for picking this one up for me! I can't approve it since it's my own PR, but I think this should fully cover the issue. 👍

**Comment by rowan-m** [comment]
Just chiming in with another successful install here. Thanks!

**Comment by sirugh** [comment]
> Just chiming in with another successful install here. Thanks!

@rowan-m what OS? I'm still having problems on Windows :(

**Comment by rowan-m** [comment]
> @rowan-m what OS? I'm still having problems on Windows :(

Linux container on ChromeOS, so unfortunately not relevant to your situation. However, you might try:
```
yarn install --ignore-optional
```
Just to be sure you're skipping the native deps.

**Comment by syedsaleem086** [comment]
Hi,

Taken the latest pull and followed above hints still not working

image

**Comment by zetlen** [comment]
@syedsaleem086 That's strange--are you sure you have the most recent version? Make sure your yarn.lock is reset to HEAD.

**Comment by sirugh** [comment]
We are merging this now as it resolves the problem with installation of pwa-studio. If you have problems building or running local development see #898/#899.

**Comment by syedsaleem086** [comment]
> @syedsaleem086 That's strange--are you sure you have the most recent version? Make sure your yarn.lock is reset to HEAD.

Hi @zetlen ,
I have taken a pull again and did reset the yarn.lock. still no luck :(

Best Regards,
Syed

**Pull request merged**